0000-prelude-2024-future.md

```diff
@@ -0,0 +1,94 @@
+- Feature Name:  `prelude_2024_future`
+- Start Date:  2023-10-05
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC describes the inclusion of the `Future` and `IntoFuture` traits in the 2024 edition prelude.
+
+# Motivation
+[motivation]: #motivation
+
+When an `async fn` is desugared we obtain an anonymous type with a signature of `impl Future`. In order to use this type we must first be able to _name_ it, which can only be done if the  `Future` trait in scope. Currently this can be a little inconvenient since it requires `std::future::Future` to be manually imported.
+
+Most other reifications of control-flow effects have their respective types and traits included in the prelude header. To support iteration we include both the `Iterator` and `IntoIterator` traits in the prelude. To support fallibility we include both `Option` and `Result` in the prelude. This RFC proposes we include both `Future` and `IntoFuture` to match.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Let's say someone wrote an async function which takes a future and operates on it. For example, it could be something like this:
+
+```rust
+use std::future::IntoFuture;
+
+async fn meow(phrase: impl IntoFuture<Output = String>) {
+    println!("{}, meow", phrase.await);
+}
+```
+
+In the Rust 2021 edition this code would require an explicit import of the `IntoFuture` trait. When migrating to the 2024 edition the import of the `IntoFuture` trait would be taken care of by the prelude, and the code could remove the explicit import:
```

Comment:
This doesn't feel like the strongest example. Generally I find that traits are most useful in preludes when you want to implicitly use their methods. Are there applications like that?

Reply:
is this more similar to having IntoIterator in the prelude?

Reply:
Yeah, exactly. Being able to always call `.into_iter()` is really useful. Being able to return a `impl IntoIterator` or create a `Box<dyn IntoIterator>` is moderately useful, but nowhere near as valuable, both in terms of frequency and beginner-accessiblity.

Reply:
Looks like `poll` is the [only method](https://doc.rust-lang.org/beta/std/future/trait.Future.html). Hmm, not actually sure which one is more useful to show off. I guess you could do both.

Reply:
@alice-i-cecile that sounds like an argument for #1880 (inherent trait impls) instead of adding more symbols to the prelude.